app/uk/gov/hmrc/selfservicetimetopay/controllers/CalculatorController.scala

```diff
@@ -18,94 +18,302 @@ package uk.gov.hmrc.selfservicetimetopay.controllers
+
+      case other => getRedirectionDestination(other)
+
+    })
   }
 
+
   def paymentTodaySubmit: Action[AnyContent] = Action.async { implicit request =>
-    val form = createPaymentTodayForm.bindFromRequest()
-    if (form.hasErrors) {
-      Future.successful(Ok(payment_today_form.render(form, request)))
-    } else {
-      Future.successful(Redirect(routes.CalculatorController.calculateInstalmentsPresent()))
-    }
+
+    Future.successful(getKeystoreData match {
+
+      case (_, amountsDue:Some[CalculatorAmountsDue], _, _, _) => {
+        val form = createPaymentTodayForm(amountsDue.get.total).bindFromRequest()
+        if (form.hasErrors) {
+          Ok(payment_today_form.render(form, request))
+        } else {
+          Redirect(routes.CalculatorController.calculateInstalmentsPresent(None)).addingToSession(
+            "CalculatorPaymentToday" -> JacksonMapper.writeValueAsString(form.get)
+          )
+        }
+      }
+
+      case other => getRedirectionDestination(other)
+    })
   }
 
-  def calculateInstalmentsPresent: Action[AnyContent] = Action.async { implicit request =>
-    //@TODO replace with keysotre
-    val calculatorDuration:CalculatorDuration = CalculatorDuration(request.session.get("CalculatorDuration").map(_.toInt)
-      .getOrElse(paymentSchedules.minBy(_.instalments.length).instalments.length))
-    val form = createDurationForm(paymentSchedules.minBy(_.instalments.length).instalments.length,
-      paymentSchedules.maxBy(_.instalments.length).instalments.length).fill(calculatorDuration)
+  def calculateInstalmentsPresent(monthsOption:Option[String]): Action[AnyContent] = Action.async { implicit request =>
+
+    Future.successful(getKeystoreData match {
+
+      case (_, Some(amountsDue: CalculatorAmountsDue), Some(paymentToday: CalculatorPaymentToday),
+      durationOption: Option[CalculatorDuration], schedulesOption: Option[List[CalculatorPaymentSchedule]]) => {
+        schedulesOption match {
+
+          case Some(schedules: Seq[CalculatorPaymentSchedule]) => {
+            val instalmentOptionsAscending = schedules.map(_.instalments.length).sorted
+            monthsOption match {
 
-    Future.successful(Ok(calculate_instalments_form.render(paymentSchedules.filter(_.instalments.length == calculatorDuration.months).head, form,
-      paymentSchedules.map(_.instalments.length).sorted, request)))
+              case Some(monthsString: String) => {
+                try {
+                  val duration = CalculatorDuration(monthsString.toInt)
+                  val durationForm = createDurationForm(instalmentOptionsAscending.head, instalmentOptionsAscending.last).bindFromRequest()
+                  if (durationForm.hasErrors) {
+                    val months = durationOption match {
+
+                      case Some(d: CalculatorDuration) => d.months
+                      case None => durationOption match {
+
+                        case Some(duration: CalculatorDuration) => duration.months
```

Comment:
^^ this line is dead code. Did you mean to match on the same value twice once in original matches block?